spacetemplate/repository.go

```diff
@@ -0,0 +1,273 @@
+package spacetemplate
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/almighty/almighty-core/errors"
+	"github.com/almighty/almighty-core/gormsupport"
+	"github.com/almighty/almighty-core/log"
+	"github.com/almighty/almighty-core/space"
+	"github.com/almighty/almighty-core/workitem"
+	"github.com/almighty/almighty-core/workitem/link"
+	"github.com/goadesign/goa"
+	"github.com/jinzhu/gorm"
+	errs "github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
+	"golang.org/x/net/context"
+)
+
+// Repository describes interactions with space templates
+type Repository interface {
+	Create(ctx context.Context, st *SpaceTemplate) error
+	List(ctx context.Context) ([]SpaceTemplate, error)
+	Load(ctx context.Context, spaceTemplateID uuid.UUID) (*SpaceTemplate, error)
+	// Apply tries to load the space template id as stored in the space and then
+	// creates all the artifacts like work item type and work item link types
+	Apply(ctx context.Context, spaceTemplateID uuid.UUID, spaceID uuid.UUID) error
+}
+
+// NewRepository creates a new space template repository
+func NewRepository(db *gorm.DB) Repository {
+	return &GormRepository{db: db}
+}
+
+// GormRepository is the implementation of the repository interface for space
+// templates.
+type GormRepository struct {
+	db *gorm.DB
+}
+
+// Create creates a new space template.
+func (r *GormRepository) Create(ctx context.Context, st *SpaceTemplate) error {
+	defer goa.MeasureSince([]string{"goa", "db", "space_template", "create"}, time.Now())
+
+	if uuid.Equal(st.ID, uuid.Nil) {
+		st.ID = uuid.NewV4()
+	}
+
+	if err := st.Validate(); err != nil {
+		log.Info(ctx, map[string]interface{}{"space_template": st, "err": err}, "space template is invalid")
+		return errs.Wrap(err, "space template is invalid")
+	}
+
+	err := r.db.Create(st).Error
+	if err != nil {
+		goa.LogError(ctx, "error adding space template", "error", err.Error())
```

Comment:
Fixed in b92e5f24cb5a42d8a775994b73a5ad8666f3dd55. `goa.LogError` is actually still used in

./area/area.go
./codebase/codebase.go
./controller/userspace.go

And my code was inspired by the area code.